core/runtime/snapshot.rs

```diff
@@ -0,0 +1,577 @@
+  );
+
+  // Get the extensions for a second pass if we want to warm up the snapshot.
+  let warmup_exts = warmup_script.map(|_| {
+    create_snapshot_options
+      .extensions
+      .iter()
+      .map(|e| e.for_warmup())
+      .collect::<Vec<_>>()
+  });
+
+  let mut js_runtime = JsRuntimeForSnapshot::new(RuntimeOptions {
+    startup_snapshot: create_snapshot_options.startup_snapshot,
+    extensions: create_snapshot_options.extensions,
+    skip_op_registration: create_snapshot_options.skip_op_registration,
+    ..Default::default()
+  });
+
+  println!("JsRuntimeForSnapshot prepared, took {:#?}", mark.elapsed(),);
+  mark = Instant::now();
+
+  let files_loaded_during_snapshot = js_runtime
+    .files_loaded_from_fs_during_snapshot()
+    .iter()
+    .map(PathBuf::from)
+    .collect::<Vec<_>>();
+
+  if let Some(with_runtime_cb) = create_snapshot_options.with_runtime_cb {
+    with_runtime_cb(&mut js_runtime);
+  }
+
+  let mut snapshot = js_runtime.snapshot();
+  if let Some(warmup_script) = warmup_script {
+    let warmup_exts = warmup_exts.unwrap();
+
+    // Warm up the snapshot bootstrap.
+    //
+    // - Create a new isolate with cold snapshot blob.
+    // - Run warmup script in new context.
+    // - Serialize the new context into a new snapshot blob.
+    let mut js_runtime = JsRuntimeForSnapshot::new(RuntimeOptions {
+      startup_snapshot: Some(Snapshot::JustCreated(snapshot)),
+      extensions: warmup_exts,
+      skip_op_registration: true,
+      ..Default::default()
+    });
+    js_runtime.execute_script_static("warmup", warmup_script)?;
+
+    snapshot = js_runtime.snapshot();
+  }
+
+  println!(
+    "Snapshot size: {}, took {:#?}",
+    snapshot.v8_len(),
+    mark.elapsed(),
+  );
+  mark = Instant::now();
+
+  let output = snapshot.serialize(create_snapshot_options.serializer)?;
+
+  println!(
+    "Snapshot written, took: {:#?}",
+    Instant::now().saturating_duration_since(mark),
+  );
+
+  Ok(CreateSnapshotOutput {
+    files_loaded_during_snapshot,
+    output,
+  })
+}
+
+pub type FilterFn = Box<dyn Fn(&PathBuf) -> bool>;
+
+pub fn get_js_files(
+  cargo_manifest_dir: &'static str,
+  directory: &str,
+  filter: Option<FilterFn>,
+) -> Vec<PathBuf> {
+  let manifest_dir = Path::new(cargo_manifest_dir);
+  let mut js_files = std::fs::read_dir(directory)
+    .unwrap()
+    .map(|dir_entry| {
+      let file = dir_entry.unwrap();
+      manifest_dir.join(file.path())
+    })
+    .filter(|path| {
+      path.extension().unwrap_or_default() == "js"
+        && filter.as_ref().map(|filter| filter(path)).unwrap_or(true)
+    })
+    .collect::<Vec<PathBuf>>();
+  js_files.sort();
+  js_files
+}
+
+pub(crate) struct SnapshottedData {
+  pub module_map_data: ModuleMapSnapshotData,
+  pub js_handled_promise_rejection_cb: Option<v8::Global<v8::Function>>,
+  pub ext_source_maps: HashMap<String, Vec<u8>>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub(crate) struct RawSnapshottedData {
```

Comment:
Not a fan of this name - I don't see what's "raw" about this data - maybe `AuxSnapshotData`?

Reply:
I can rename this

Reply:
This is now `SerializableSnapshotSidecarData` to indicate that its intended destination is serde.